### ISSUE TITLE
Types converted to "any" in JavaScript

### ISSUE THREAD

**Issue by TheColorRed** [Microsoft/TypeScript#21362]
I am writing a library, and I have two folders, one with my source code, and one with my tests.

The tests are written in javascript while the source is written in typescript.

I have the following `jsconfig.json` located here `public/app/jsconfig.json`:

```json
{
  "typeAcquisition": {
    "include": [
      "../../lib/hp.d.ts"
    ]
  }
}
```

In the `.d.ts` file I have multiple methods, for example I have this one:

```ts
parentComponent<T extends component>(comp: ComponentType<T>, callback?: (component: T) => void): T | null;
```

When I use that method in the javascript (vscode), It is converting items to `any` like this:

```ts
parentComponent<any>(comp: ComponentType<any>, callback?: (component: any) => void): any
```

This is removing items from intellisense. Is this supposed to be happening?

**Comment by andy-ms** [comment]
Are you talking about signature help? If you haven't written any arguments yet we can't infer the type parameter, so it becomes `any`. However, if you have `--checkJs` we should give you errors if you use it incorrectly.

**Comment by TheColorRed** [comment]
Take this javascript:

```js
class B extends hp.element {}

class A extends hp.element {
  created() {
    this.parentComponent(B, item => {
      item.// Show suggestions
    })
  }
}
```

In the callback the `.d.ts` says that `item` will be an instance of `B`, however `item` is `any`.

**Comment by mhegazy** [comment]
Is `hb.element` generic? if so, generic type parameters that are not specified will default to `any` in a .js file. use [`@augments`](https://github.com/Microsoft/TypeScript/wiki/Type-Checking-JavaScript-Files#unspecified-type-parameters-default-to-any) to specify these.

**Comment by TheColorRed** [comment]
in the TypeScript `hp.element` looks like this:

```ts
namespace hp {
  export abstract class element extends component {
  }
}
```

So... I believe to answer your question... no it isn't.

**Comment by andy-ms** [comment]
Could you provide a complete example that reproduces this problem?

**Comment by TheColorRed** [comment]
**Github Repo:** https://github.com/TheColorRed/horsepower
**Compiled library files (.js and .d.ts):** [lib.zip](https://github.com/Microsoft/TypeScript/files/1661608/lib.zip)

Example Class

```js
class stopwatchwrapper extends hp.element {}
class stopwatch extends hp.element {}

class stopwatchbutton extends hp.button {
  clicked() {
    this.closestComponent(stopwatchwrapper, item => {
      // item is `any`
      item.childComponent(stopwatch, item => {
        // item is `any`
      })
    })
  }
}
```

**Comment by andy-ms** [comment]
@TheColorRed That is a TypeScript project -- there's no single `.d.ts` file there for javascript to reference. You are probably missing types for some things -- turn on `// @ts-check` to see. When I added the following I got non-`any` callbacks:
```ts
// @ts-check

/// <reference path="core/components/dom/element.ts" />
/// <reference path="core/components/dom/elements/button.ts" />
/// <reference path="core/component.ts" />
/// <reference path="core/components/dom/form.ts" />
```

**Comment by TheColorRed** [comment]
The project needs to be built and the `.d.ts` will be created, that is why I provided the zip file so you don't need to build it.

You can use that `.d.ts` file and reference it.

**Comment by andy-ms** [comment]
If I build the repo, then `/// <reference path="hp.d.ts" />`, I get a typed `item` too.

**Comment by TheColorRed** [comment]
As you can see I don't get that `public/app/stopwatch.js`:

![non-typed](https://user-images.githubusercontent.com/1571806/35360624-00335176-0124-11e8-926b-1ed2d4ddff7a.png)

**Comment by andy-ms** [comment]
@TheColorRed Can you try and reproduce the error in a sample project with `checkJs`? There are a number of reasons it might not be working for you and it's hard to tell from a screenshot.

**Comment by TheColorRed** [comment]
Where do I put `checkJs`? I have it in the `jsconfig.json` that is located in `public/app`

**Comment by andy-ms** [comment]
Try with a `tsconfig.json`, and try compiling with `tsc` on the command-line instead of using an IDE. All you should need is this:
```json
{
	"compilerOptions": {
		"allowJs": true,
		"checkJs": true
	}
}
```
and `a.d.ts` and `b.js` in the same directory.

**Comment by TheColorRed** [comment]
I am still getting `any` as a parameter type

```json
{
  "compilerOptions": {
    "allowJs": true,
    "checkJs": true,
    "module": "system",
    "target": "es2015",
    "outFile": "lib/hp.js",
    "declaration": true,
    "strict": true,
    "noUnusedLocals": true,
    "noImplicitAny": true,
    "lib": [
      "dom",
      "dom.iterable",
      "es2015"
    ]
  },
  "include": [
    "src/util/**/*.ts",
    "src/component.ts",
    "src/element.ts",
    "src/core/**/*.ts",
    "src/index.ts"
  ]
}
```

**Comment by mhegazy** [comment]
What version of VSCode and TS are you using?

**Comment by TheColorRed** [comment]
My fault, I feel stupid for not testing with `typescript@next` before reporting the issue. Installing that fixes my issue.

Sorry for wasting your time.
